oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProvider.java

```diff
@@ -66,45 +69,64 @@
  * @see OAuth2RefreshTokenAuthenticationToken
  * @see OAuth2AccessTokenAuthenticationToken
  * @see OAuth2AuthorizationService
- * @see JwtEncoder
- * @see OAuth2TokenCustomizer
- * @see JwtEncodingContext
- * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-1.5">Section 1.5 Refresh Token Grant</a>
- * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-6">Section 6 Refreshing an Access Token</a>
+ * @see OAuth2TokenGenerator
+ * @see <a target="_blank" href="https://datatracker.ietf.org/doc/html/rfc6749#section-1.5">Section 1.5 Refresh Token Grant</a>
+ * @see <a target="_blank" href="https://datatracker.ietf.org/doc/html/rfc6749#section-6">Section 6 Refreshing an Access Token</a>
  */
 public final class OAuth2RefreshTokenAuthenticationProvider implements AuthenticationProvider {
+	private static final String ERROR_URI = "https://datatracker.ietf.org/doc/html/rfc6749#section-5.2";
 	private static final OAuth2TokenType ID_TOKEN_TOKEN_TYPE = new OAuth2TokenType(OidcParameterNames.ID_TOKEN);
 	private static final StringKeyGenerator DEFAULT_REFRESH_TOKEN_GENERATOR =
 			new Base64StringKeyGenerator(Base64.getUrlEncoder().withoutPadding(), 96);
 	private final OAuth2AuthorizationService authorizationService;
-	private final JwtEncoder jwtEncoder;
-	private OAuth2TokenCustomizer<JwtEncodingContext> jwtCustomizer = (context) -> {};
+	private final OAuth2TokenGenerator<? extends OAuth2Token> tokenGenerator;
 	private Supplier<String> refreshTokenGenerator = DEFAULT_REFRESH_TOKEN_GENERATOR::generateKey;
```

Comment:
If we eventually have refresh tokens produced by `OAuth2TokenGenerator` as well, would the plan be to deprecate the `refreshTokenGenerator` field and use a single generator for all tokens? In this case, would the default implementation be a delegation pattern that encapsulates a `JwtGenerator` and `OpaqueTokenGenerator`, delegated to based on token type? Or will there be a second `OAuth2TokenGenerator` just for refresh tokens?

I'm just trying to visualize how we will have a combination of `Jwt` and non-`Jwt` generators in one class.

Reply:
> would the plan be to deprecate the `refreshTokenGenerator` field and use a single generator for all tokens

Yes. We would leverage a delegation-based pattern where `DelegatingOAuth2TokenGenerator` would be composed of `JwtGenerator` and `OAuth2OpaqueTokenGenerator`. The `OAuth2OpaqueTokenGenerator` would contain pre-processing logic as such:

```java
public class OAuth2OpaqueTokenGenerator implements OAuth2TokenGenerator<OAuth2OpaqueToken> {

	@Override
	public OAuth2OpaqueToken generate(OAuth2TokenContext context) {
		if (OAuth2TokenType.ACCESS_TOKEN.equals(context.getTokenType()) &&
				context.getRegisteredClient().getTokenSettings().getAccessTokenType().isOpaque()) {

			// Generate opaque access token

		}

		if (OAuth2TokenType.REFRESH_TOKEN.equals(context.getTokenType())) {

			// Generate opaque refresh token

		}
```

We will need to introduce a setting in `TokenSettings` that allows us to determine which token type format to generate for an access token (jwt or opaque) and we'll default to opaque for refresh tokens.